### ISSUE TITLE
GDPR Review

### ISSUE THREAD

**Issue by joeaudette** [cloudscribe/cloudscribe#451]
We've had a cookie consent solution for a while now, but in asp.net core 2.1 Microsoft added some new gdpr features including cookie consent, and the ability for users to download their personal data and delete personal data.

We should review our cookie consent solution and see if the one from Microsoft might be better.
We should add the download and delete capabilities for personal user data.

Sql Azure databases are now encrypted by default. Those using Sql Server or other db platforms that support database encryption should consider using database encryption to help meet GDPR requirements.

**Comment by joeaudette** [comment]
After initial review I think we should:

* Replace our cookie consent with the one provided by Microsoft.
* Add a built in privacy policy page that we can link to from the consent partial view
* Add functionality for users to delete their account and personal data
* Add functionality so users can download their personal data
* Add some kind of interfaces so that custom features may provide links to download personal data captured by custom features, and can if required provide a link to delete personal data captured by custom features.
* Add a built in dedicated page for users to manage their personal data, this is where the interfaces would plugin the additional links for feature specific data. This way there is a central location for a user to manage all the data captured by any custom feature. Of course it is up to the custom feature to implement the interfaces.

Keeping in mind that some data captured by custom features may not be required to let the user delete the data. Businesses have a legal requirement to keep data related to business transactions, so for example ecommerce features do not have enable users to delete their order history as far as I know, but they do have to make it downloadable as far as I know.

Implementation challenges for multi-tenancy

I think we will need a per tenant CookiPolicyOptionsMonitor so that each tenant can enable or disable CheckConsentNeeded, the default implementation would be global to the app.

**Comment by joeaudette** [comment]
I've implemented the ideas mentioned in my previous post but am still reviewing the broader impact of GDPR.

Some additional info as I've been working on improving our built in tools for helping with GDPR compliance.

Privacy policies should also explain about any 3rd party cookies and 3rd party cookie should be prevented by not loading 3rd party scripts (that in turn add their own cookies) if we don't have cookie consent.

This impacts even the use of such things as recaptcha
https://law.stackexchange.com/questions/27908/gdpr-recaptcha-with-users-consent
https://www.google.com/about/company/user-consent-policy.html

and of course to google analytics, AddThis, Disqus and the like

**Comment by joeaudette** [comment]
From reading the above, there is a reasonable argument that recaptcha would meet the exclusions requirements as mentioned in the answer above from stackexchenge because it "is necessary and proportionate for the purposes of ensuring network and information security"

**Comment by joeaudette** [comment]
I've published updated nugets and updated project templates that sets up the new GDPR features mentioned above. Users can download their personal data and delete their account if they wish.

Without cookie consent, AddThis scripts are not added by default nor is gravatar used since both add 3rd party cookies. google analytics is anonymized and I understand that to be sufficient as long as not using the advertising features of GA.

In SimpleContent I made it not load the disqus scripts for comments without cookie consent as well, instead it shows a message that to view or add comments requires cookie consent.

After all this work I think there is still one more thing to do that they did not do in the default asp.net mvc template. Once a user accepts cookies, ie gives consent, they must also be provided a way to revoke that consent. I will implement something for that soon.
